In [1]:
%env KERAS_BACKEND=tensorflow

env: KERAS_BACKEND=tensorflow


In [2]:
%matplotlib inline

import numpy as np
import matplotlib.pyplot as plt

讀入 IMDB 電影數據庫

In [3]:
from keras.datasets import imdb

C:\Users\user\Anaconda3\lib\site-packages\h5py\__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.


In [4]:
(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=10000)

處理輸入資料

In [5]:
from keras.preprocessing import sequence

In [6]:
x_train = sequence.pad_sequences(x_train, maxlen=150)
x_test = sequence.pad_sequences(x_test, maxlen=150)

In [7]:
x_train.shape

(25000, 150)

建構RNN模型
嘗試將輸入壓成120維度，而神經元數設30個

In [82]:
N=120
K=30

In [54]:
from keras.models import Sequential
from keras.layers import Dense, Embedding
from keras.layers import LSTM

In [83]:
model = Sequential()

In [84]:
model.add(Embedding(10000, N))

In [85]:
model.add(LSTM(K))

In [86]:
model.add(Dense(1, activation='sigmoid'))

組裝模型

In [87]:
model.compile(loss='binary_crossentropy',
             optimizer='adam',
             metrics=['accuracy'])

In [88]:
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_8 (Embedding)      (None, None, 120)         1200000   
_________________________________________________________________
lstm_8 (LSTM)                (None, 30)                18120     
_________________________________________________________________
dense_8 (Dense)              (None, 1)                 31        
Total params: 1,218,151
Trainable params: 1,218,151
Non-trainable params: 0
_________________________________________________________________


訓練

In [89]:
model.fit(x_train, y_train,
         batch_size=32,
         epochs=10,
         validation_data = (x_test, y_test))

Train on 25000 samples, validate on 25000 samples
Epoch 1/10
25000/25000 [==============================] - 116s 5ms/step - loss: 0.4015 - acc: 0.8155 - val_loss: 0.3116 - val_acc: 0.8659
Epoch 2/10
25000/25000 [==============================] - 114s 5ms/step - loss: 0.2338 - acc: 0.9070 - val_loss: 0.3163 - val_acc: 0.8659
Epoch 3/10
25000/25000 [==============================] - 116s 5ms/step - loss: 0.1733 - acc: 0.9334 - val_loss: 0.3559 - val_acc: 0.8604
Epoch 4/10
25000/25000 [==============================] - 106s 4ms/step - loss: 0.1254 - acc: 0.9553 - val_loss: 0.3947 - val_acc: 0.8514
Epoch 5/10
25000/25000 [==============================] - 105s 4ms/step - loss: 0.0956 - acc: 0.9664 - val_loss: 0.4846 - val_acc: 0.8542
Epoch 6/10
25000/25000 [==============================] - 105s 4ms/step - loss: 0.0863 - acc: 0.9706 - val_loss: 0.4688 - val_acc: 0.8546
Epoch 7/10
25000/25000 [==============================] - 103s 4ms/step - loss: 0.0583 - acc: 0.9802 - val_loss: 0.5564 - 

In [90]:
score = model.evaluate(x_test, y_test)

25000/25000 [==============================] - 18s 711us/step


In [92]:
print(f'測試資料的 loss = {score[0]}')
print(f'測試資正確率 = {score[1]}')

測試資料的 loss = 0.6247740922403335
測試資正確率 = 0.85236


由於訓練資料的正確率成長和測試資料的正確率持長不成正比，經過多次修改N(維度)和K(神經元數)測試後，最高的正確率似乎只能停留在0.85左右